List of parameters in the stack

Marwan99 opened the pull request:
This should contain a list of all the declared parameters in the stack #1589.

Was thinking of having each parameter in it's package readme, let me know what you think of the best format and location for this data.

SteveMacenski commented:
Super good stuff - can you go over really quick how you compiled this list? This will help me make sure we got everything and in the future as new things are made making sure we keep this up to date. Also, #1763 brings up that there are some params that have been `get` but not `declared`, which is annoying. 

Things that jump out to me that may be missing:
- The controller server I see minimum params but not maximum params for velocities
- Static layer doesnt have a topic param for map
- There's hyperlinks that just redirect to this page - I think you're trying to represent a namespace in a way that is incorrect. 
- Voxel layer also has all the same params as obstacle layer + those new ones since it derives from it
- the `dwb_plugin_name_.PathAlign.scale` the `<dwb controller name>.<critic plugin name>.scale`, you have the default plugin critics listed with their scale/weight, but those are actually for all the critic plugins
- What's `map_grid`?
- Oscillation looks like it might have more parameters
- Planner server looks to be missing params potentially
- Does backup not have a speed param?

Agreed on figuring out where these should live. I think the answer is the website but how we sort them is a good question. I think for now the markdown file is perfect because I can just sit down and fill in descriptions in one place.

SteveMacenski commented:
Also look at this list if there's anything missing from yours this has: https://navigation.ros.org/configuration/params/tunable-params.html its a very old list and it wasn't complete at the time

Marwan99 commented:
I  mainly searched for `declare`, here is a list of functions/methods I have found so far used for declaring parameters:
*  `rclcpp_lifecycle::LifecycleNode::declare_parameter`
*  `nav2_costmap_2d::Layer::declareParameter`  
* `nav2_util::declare_parameter_if_not_declared`
* `nav2_util::LifecycleNode::add_parameter`
*  `nav_2d_utils::searchAndGetParam` (noticed this one after you have mentioned oscillations are missing params)

#### Addressing comments:

* controller server doesnt seem to have max velocities: https://github.com/ros-planning/navigation2/blob/7c0ac74514506591c0d508a6bc10e42d1d67162f/nav2_controller/src/nav2_controller.cpp#L39-L48
* `map_topic` is declared in costmap_2d_ros.cpp, and so in theory the static layer doesn't need to re-declare it, is that correct?
* This is [map_grid](https://github.com/ros-planning/navigation2/blob/master/nav2_dwb_controller/dwb_critics/src/map_grid.cpp), added TrajectoryCritic beside it in the previous commit.
* planner server looks complete to me, did you expect something and didn't see it?
* backup seems to be getting the speed and distance from here
https://github.com/ros-planning/navigation2/blob/7c0ac74514506591c0d508a6bc10e42d1d67162f/nav2_behavior_tree/plugins/action/back_up_action.cpp#L63-L64
* The remaining comments should be have been addressed in the previous commit.

SteveMacenski commented:
@Marwan99 awesome, thanks for the clarification - I agree on all points. 

I think then the last thing then is the inport ports for the BT nodes. We can create a separated section `---` at the bottom for BT port parameters. Something like `BT Node Name (BT Ports)` to denote that they're ports. I think that will then be a complete list to work from.

At that point, I think then the path is to merge this, I assign myself to write the descriptions, and merge those changes in. Afterwards, I'm not exactly what the best way is to move that documentation to the website, but I'm sure we'll figure it out. Any suggestions would be good. Maybe replace the dropdowns here (https://navigation.ros.org/configuration/params/tunable-params.html) for our updated tables with descriptions. Or make something more looking like the controller page (https://navigation.ros.org/configuration/configuring-the-controller.html)?

SteveMacenski commented on doc/parameters/param_list.md:
```diff
@@ -420,4 +420,121 @@ Namespace: /parent_ns/local_ns
 | z_short | 0.05 | |
-| always_reset_initial_pose | false | Requires that AMCL is provided an initial pose either via topic or initial_pose* parameter (with parameter set_initial_pose: true) when reset. Otherwise, by default AMCL will use the last known pose to initialize | --!>
\ No newline at end of file
+| always_reset_initial_pose | false | Requires that AMCL is provided an initial pose either via topic or initial_pose* parameter (with parameter set_initial_pose: true) when reset. Otherwise, by default AMCL will use the last known pose to initialize | --!>
```
Why the `--!>`?

Marwan99 replied:
That has been fixed in 7348daf

Marwan99 commented:
I am not sure if there is a default `server_name` for the action BT nodes, I have done some digging but  I couldn't find something, @SteveMacenski please let me know if I missed it.

>Maybe replace the dropdowns here (https://navigation.ros.org/configuration/params/tunable-params.html) for our updated tables with descriptions. Or make something more looking like the controller page

I like the idea of replacing the dropdown lists with tables as I currently struggle with searching for things via `CTRL+F`. We may also have hyperlinks at the top of  the page to allow for quickly jumping to a specific node. Having dedicated pages like the controller page sounds cool as well.

SteveMacenski commented:
There is no default, you would be correct. This is used so that if its not provided, we default to our usual names. This is mostly useful for namespacing in case you have multiple planner servers for some reason. 

I think in the end of this work, we'd have a single page with all of the parameters (maybe split into sub-pages, depending on length and formatting) and then we'd have a separate configuration guide section that would cover the controller amongst other things. 

Something I just noticed is that sometimes you use `source.`, `voxel_layer.` `dwb_plugin_name_.` or no namespace at all for parameters that are stuck under a specific namespace set by the name/ID. I think we should make this consistent somehow. 

My suggestion would be `<voxel layer>.` `<dwb plugin>.`, etc with the pattern of `<>` surrounding the base name `voxel layer` `obstacle layer` `dwb controller` and then have a note at top saying something along the lines of "NOTE: `<>` means plugin are namespaced by a name/id parameterization. The bracketed names may change due to your application configuration"

I can start working on the parameter descriptions then. If you had thoughts about how to put this on the website in the dropdowns cleanly (as tables or otherwise) I'd be open to seeing a suggestion. I'd just start with 1 section that only has a few to get the idea across and not take too long. I kind of like the drop downs of this: https://navigation.ros.org/configuration/params/tunable-params.html but with the content of this https://navigation.ros.org/configuration/configuring-the-controller.html within the drop down (so not just a YAML renderer, but a little table for each param). 

Another example that may inspire you: http://ceres-solver.org/nnls_solving.html#_CPPv2N5ceres6Solver7Options33gradient_check_relative_precisionE

Marwan99 commented:
@SteveMacenski what do you think of the notes explaining the base names?

I will have a go at the website over the weekend and submit the sample as a PR then.

SteveMacenski commented:
Absolutely perfect, thank you. I like your added note at the start of each section too, that will be really helpful for lazy readers. I see you filed a ticket #1783 should that block this?
